core/atomic_c11.c

```diff
@@ -0,0 +1,363 @@
+TEMPLATE_ATOMIC_OP_FETCH_N( and, &, 8,  ) /* __atomic_and_fetch_8 */
+
+TEMPLATE_ATOMIC_OP_FETCH_N(  or, |, 1,  ) /* __atomic_or_fetch_1 */
+TEMPLATE_ATOMIC_OP_FETCH_N(  or, |, 2,  ) /* __atomic_or_fetch_2 */
+TEMPLATE_ATOMIC_OP_FETCH_N(  or, |, 4,  ) /* __atomic_or_fetch_4 */
+TEMPLATE_ATOMIC_OP_FETCH_N(  or, |, 8,  ) /* __atomic_or_fetch_8 */
+
+TEMPLATE_ATOMIC_OP_FETCH_N( xor, ^, 1,  ) /* __atomic_xor_fetch_1 */
+TEMPLATE_ATOMIC_OP_FETCH_N( xor, ^, 2,  ) /* __atomic_xor_fetch_2 */
+TEMPLATE_ATOMIC_OP_FETCH_N( xor, ^, 4,  ) /* __atomic_xor_fetch_4 */
+TEMPLATE_ATOMIC_OP_FETCH_N( xor, ^, 8,  ) /* __atomic_xor_fetch_8 */
+
+TEMPLATE_ATOMIC_OP_FETCH_N(nand, &, 1, ~) /* __atomic_nand_fetch_1 */
+TEMPLATE_ATOMIC_OP_FETCH_N(nand, &, 2, ~) /* __atomic_nand_fetch_2 */
+TEMPLATE_ATOMIC_OP_FETCH_N(nand, &, 4, ~) /* __atomic_nand_fetch_4 */
+TEMPLATE_ATOMIC_OP_FETCH_N(nand, &, 8, ~) /* __atomic_nand_fetch_8 */
+
+/* ***** Generic versions below ***** */
+
+/* Clang objects if you redefine a builtin.  This little hack allows us to
+ * define a function with the same name as an intrinsic. */
+/* Hack origin: http://llvm.org/svn/llvm-project/compiler-rt/trunk/lib/builtins/atomic.c */
+#pragma redefine_extname __atomic_load_c __atomic_load
+#pragma redefine_extname __atomic_store_c __atomic_store
+#pragma redefine_extname __atomic_exchange_c __atomic_exchange
+#pragma redefine_extname __atomic_compare_exchange_c __atomic_compare_exchange
+
+/**
+ * @brief Atomic generic load
+ *
+ * @param[in]  size       width of the data, in bytes
+ * @param[in]  src        source address to load from
+ * @param[in]  dest       destination address
+ * @param[in]  memorder   memory ordering, ignored in this implementation
+ */
+void __atomic_load_c(size_t size, const void *src, void *dest, int memorder)
+{
+    (void) memorder;
+    unsigned int mask = irq_disable();
+    memcpy(dest, src, size);
+    irq_restore(mask);
+}
+
+/**
+ * @brief Atomic generic store
+ *
+ * @param[in]  size       width of the data, in bytes
+ * @param[in]  dest       destination address to store to
+ * @param[in]  src        source address
+ * @param[in]  memorder   memory ordering, ignored in this implementation
+ */
+void __atomic_store_c(size_t size, void *dest, const void *src, int memorder)
+{
+    (void) memorder;
+    unsigned int mask = irq_disable();
+    memcpy(dest, src, size);
+    irq_restore(mask);
+}
+
+/**
+ * @brief Atomic generic exchange
+ *
+ * @param[in]  size       width of the data, in bytes
+ * @param[in]  ptr        object to swap
+ * @param[in]  val        value to swap to
+ * @param[in]  ret        put the old value from @p ptr in @p ret
+ * @param[in]  memorder   memory ordering, ignored in this implementation
+ */
+void __atomic_exchange_c(size_t size, void *ptr, void *val, void *ret, int memorder)
+{
+    (void) memorder;
+    unsigned int mask = irq_disable();
+    memcpy(ret, ptr, size);
+    memcpy(ptr, val, size);
```

Comment:
It is implemented as the manual specifies it:

https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html